### ISSUE TITLE
Issue 4: Strengthen Cross-Dataset and Cross-Model Generalization

### ISSUE THREAD

**Issue by Mati86** [AI-Security-Internships-2026/08-onnx-semantic-security-engine#24]
# Issue 4: Strengthen Cross-Dataset and Cross-Model Generalization

## Depends On: M1.1, M1.2

---

## Problem Statement

The current study centers on one lightweight MLP and a limited cross-dataset setting, while the paper describes the assurance layer as architecture-agnostic. Stronger evidence is required that the findings are not specific to one model or dataset pair.

---

### 1. Add One Additional External Dataset

Prefer standardized NetFlow datasets. Recommended minimum design:

```
Train: NF-CSE-CIC-IDS2018-v2
Test:  NF-ToN-IoT-v2
       NF-BoT-IoT-v2
```

Document exact versions and preprocessing.

### 2. Separate Two Cross-Dataset Questions

**Track A — Domain/distribution shift:** same standardized feature schema across datasets.

**Track B — Feature-extractor/semantic mismatch:** CICFlowMeter ↔ nProbe/NetFlow mapping scenario.

Do not combine both causes under one generic "domain shift" explanation.

### 3. Add One Additional Neural Model

Keep the current MLP and add one compact ONNX-compatible neural architecture with an intermediate embedding (e.g. small 1D-CNN).

Do not add many architectures.

### 4. Repeat Core Assurance Evaluation

For both neural models report ID classification, OOD performance, false-positive behavior, key failure-mode coverage and SEMANTICSHIELD overhead.

### 5. Keep Non-Neural Baselines Separate

RF/XGBoost may be classifier baselines if useful, but do not force the embedding-based method onto models where it is not naturally applicable.

---

## Acceptance Criteria

- [ ] At least one additional external NIDS dataset is evaluated
- [ ] Same-schema domain shift is separated from schema/semantic mismatch
- [ ] One additional compact neural ONNX model is evaluated
- [ ] Core assurance methodology is tested on both neural models
- [ ] Dataset versions/features/preprocessing are documented
- [ ] Binary and multiclass cross-dataset results are separated where applicable
- [ ] Architecture-generality claims are limited to demonstrated evidence
- [ ] Outputs feed the canonical paper-results pipeline

---

## Scope Guardrail

This is **not** a competition to find the best NIDS architecture. One additional model and one additional external dataset are sufficient unless they reveal a methodological problem.

---

### E2.2-A — Same-Schema Cross-Dataset Shift

Train/calibrate on the selected standardized source dataset and evaluate unchanged on at least two external standardized datasets. Report classifier performance and assurance/OOD performance separately.

### E2.2-B — Semantic / Extractor Mismatch

Using the verified mapping from M3.1 compare exact/compatible mappings, approximate mappings and known mismatches where experimentally appropriate.

### E2.2-C — Cross-Model Replication

Repeat the main ID/OOD protocol for the current MLP and one additional compact neural model using comparable splits and the same OOD datasets.

## Required Results

Report:
- ID accuracy/macro-F1 per model
- external-dataset accuracy/macro-F1
- AUROC/AUPRC per assurance monitor
- observed ID FPR
- TPR at fixed-FPR budgets
- cross-model runtime overhead
- effect of semantic mismatch on classifier and assurance behavior

### Table — Cross-Dataset Generalization

| Train Dataset | Test Dataset | Model | Accuracy | Macro-F1 | MSP AUROC | Mahalanobis AUROC | Full Assurance Metric |

### Table — Cross-Model Replication

| Model | Size/Params | ID Macro-F1 | OOD AUROC | TPR@1% FPR | Assurance Overhead |

### Figure — ID vs External-Dataset Performance

Show classifier performance drop from ID to each external dataset.

### Figure — Detector Generalization Across Datasets

Compare the primary OOD metric of each monitor across datasets.

### Table/Figure — Semantic Mapping Sensitivity

Compare exact/compatible mappings against approximate/mismatched mappings.

**Comment by Mati86** [comment]
Please ensure:

1. Evaluate at least one additional external NIDS dataset beyond the current cross-dataset evaluation. Prefer standardized NetFlow-compatible datasets where possible.
2. Clearly separate same-schema distribution/domain shift from feature-schema or semantic mismatch. These are different failure conditions and must not be reported as the same experiment.
3. Add at least one additional compact neural classifier architecture (e.g., a compact 1D-CNN) alongside the existing MLP and export/use it through ONNX.
4. Apply the same SEMANTICSHIELD assurance methodology to both neural models without redesigning the method specifically for the second model.
5. For each dataset/model combination, report relevant classifier performance (accuracy/macro-F1) separately from assurance/OOD performance (AUROC, AUPRC, observed ID FPR, and TPR at the fixed FPR operating points established in Issue #23).
6. Document dataset versions, feature sets, preprocessing/alignment, training/testing roles, model configuration, and any feature mappings required.
7. Where binary and multiclass evaluations differ, report them separately and do not directly compare incompatible metrics.
8. Generate cross-dataset and cross-model tables/figures from machine-readable outputs under the canonical paper_results pipeline.

Important: Do not claim that SEMANTICSHIELD is architecture-independent or universally generalizable. The conclusion should be limited to what is demonstrated by the tested datasets and the two neural architectures.

When complete, please comment here with the datasets used, models evaluated, experiment commands, result paths, generated tables/figures, and a concise summary of what generalized successfully and what did not.
